src/Build/Graph/ProjectGraph.cs

```diff
@@ -317,10 +330,73 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
             var graphNode = new ProjectGraphNode(
                 projectInstance,
                 globalProperties);
-            _allParsedProjects.Add(configurationMetadata, graphNode);
+            _allParsedProjects[configurationMetadata] = graphNode;
             return graphNode;
         }
 
+        /// <summary>
+        /// Load a graph with root node at entryProjectFile
+        /// Maintain a queue of projects to be processed and evaluate projects in parallel
+        /// </summary>
+        private void LoadGraph(ConcurrentQueue<ConfigurationMetadata> projectsToEvaluate, ProjectCollection projectCollection, ConcurrentDictionary<ConfigurationMetadata, object> tasksInProgress)
+        {
+            var waitHandle = new AutoResetEvent(false);
+            while (projectsToEvaluate.Count != 0 || tasksInProgress.Count() != 0)
+            {
+                ConfigurationMetadata projectToEvaluate;
+                if (projectsToEvaluate.Count != 0)
+                {
+                    projectToEvaluate = projectsToEvaluate.Dequeue();
+                    var task = new Task(() =>
+                    {
+                        ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
+                        IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.ProjectInstance.GetItems(ProjectReferenceItemName);
+                        foreach (var projectReferenceToParse in projectReferenceItems)
+                        {
+                            if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                            {
+                                throw new InvalidOperationException(string.Format(
+                                    CultureInfo.InvariantCulture,
+                                    ResourceUtilities.GetResourceString(
+                                        "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                                    projectReferenceToParse.EvaluatedInclude,
+                                    parsedProject.ProjectInstance.FullPath));
+                            }
+
+                            string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+                            PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties = GetProjectReferenceGlobalProperties(projectReferenceToParse, projectToEvaluate.GlobalProperties);
+                            var projectReferenceConfigurationMetadata = new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
+                            if (!_allParsedProjects.TryGetValue(projectReferenceConfigurationMetadata, out var _))
+                            {
+                                if (!tasksInProgress.TryGetValue(projectReferenceConfigurationMetadata, out var _))
```

Comment:
There may be a race condition here between `_allParsedProjects` and `tasksInProgress`. Independently they're thread-safe, but the state of one might change while you're checking the other.

Maybe flipping the order would be slight better even. Right now it might not be fully parsed when you check `_allParsedProjects`, then it finishes before you check `tasksInProgress`. If you flip it though, you at least wouldn't see that specific race condition. Not sure if there may be others though.

If so, you'll need a single lock around any accesses to the state required for processing (`_allParsedProjects`, `tasksInProgress` and `projectsToEvaluate`).

Reply:
This is a good catch. I _think_ just flipping the order works, but I'm not that confident.